core-java/src/main/java/eu/neverblink/jelly/core/internal/LastNodeHolder.java

```diff
@@ -1,13 +1,62 @@
 package eu.neverblink.jelly.core.internal;
 
+import eu.neverblink.jelly.core.InternalApi;
+
 /**
  * Tiny mutable holder for the last node that occurred as S, P, O, or G.
  * @param <TNode> the type of the node
  */
+@InternalApi
 public class LastNodeHolder<TNode> {
```

Comment:
I'm not excited about the changes here. The hasValue checks are only applicable to graph nodes, but we bear their cost also with subject, predicate, object nodes, which is bad (this IS a hot spot in the profiler graph).

Please revert this to how it used to be for S, P, O, and only make this more extensive version for the graph node.